zmq_actor.hpp

```diff
@@ -0,0 +1,99 @@
+    of this software and associated documentation files (the "Software"), to
+    deal in the Software without restriction, including without limitation the
+    rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+    sell copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+*/
+
+#ifndef __ZMQ_ACTOR_HPP_INCLUDED__
+#define __ZMQ_ACTOR_HPP_INCLUDED__
+
+#include "zmq.hpp"
+
+#ifdef ZMQ_CPP11
+
+#include <thread>
+
+namespace zmq
+{
+/*! Return two paired PAIR sockets.
+   
+  First is bound and second is connected via inproc://.
+*/
+std::pair<zmq::socket_t, zmq::socket_t> create_linked_pairs(context_t &ctx)
+{
+    std::pair<zmq::socket_t, zmq::socket_t> ret{socket_t(ctx, socket_type::pair),
+                                                socket_t(ctx, socket_type::pair)};
+
+    std::stringstream ss;
+    ss << "inproc://paired-" << std::hex << ret.first.handle() << "-"
+       << ret.second.handle();
+    std::string addr = ss.str();
+    ret.first.bind(addr.c_str());
+    ret.second.connect(addr.c_str());
+    return ret;
+}
+
+/*! Spawn a function in a thread and communicate over a link.
+     
+  The actor function must take a socket and zero or more optional
+  arguments such as:
+     
+      void func(zmq::socket sock, ...);
+      zmq::actor_t actor(ctx, func, ...);
+     
+  The socket passed in is one end of a bidirection link shared
+  with the application thread.  The application thread may get
+  the other end of the link by calling zmq::actor_t::link().
+     
+  Note: unlike CZMQ zactor_t, zmq::actor_t does not institute any
+  startup/shutdown protocol across the link.  If any is needed it
+  is left to the calling application.
+*/
+class actor_t
+{
+  public:
+    // Template constructor and not class to perform type erasure
+    // of the function type.
+    template<typename Func, typename... Args>
+    actor_t(context_t &ctx, Func fn, Args... args)
+    {
+        socket_t asock;
+        std::tie(asock, _sock) = create_linked_pairs(ctx);
+
+        _thread = std::thread(
+          [fn = std::forward<Func>(fn)](zmq::socket_t asock, Args... args) {
+              fn(std::move(asock), std::forward<Args>(args)...);
+          },
+          std::move(asock), std::forward<Args>(args)...);
+    }
```

Comment:
I feel like this constructor is doing too much and can be simplified
```C++
    template<class Func>
    actor_t(context_t &ctx, Func&& fn)
    {
        socket_t asock;
        std::tie(asock, _sock) = create_linked_pairs(ctx);

        _thread = std::thread(
          [sock = std::move(sock), fn = std::forward<Func>(fn)]() mutable {
              fn(std::move(asock));
              // NOTE a shutdown signal would make this more useful
          });
    }
```

Reply:
This does two unwanted things: 

1) It removes the ability to pass arguments into the function.  I guess this can be replaced by requiring the user to provide a lambda.

2) It makes the `actor_t` typed.  One use case I have is to maintain a collection of actors which may have similar protocol on the "link" but which have different actor functions.

Reply:
1. Lambdas are the way to go, but you could use `std::bind` or `std::bind_front` to construct the function object if that simplifies the code.
2. I don't think so. You can still have a `std::vector<actor_t>` with different functions. They are type erased in `std::thread`.

Reply:
About 2.  Yes, you are right!  Sorry, my lack of reading comprehension. I misunderstood the example thinking you were suggesting to template the class.  I will give both of these a try.